backend/src/rating/ranking.ts

```diff
@@ -0,0 +1,187 @@
+        dateXLikes: rawDateXLikesScore,
+        useOfChapters,
+      },
+    };
+  });
+};
+
+export const getNormalizedExternalRanking = (
+  videos: IRawYoutubeVideo[],
+  maxScores: IMaxScores
+): INormalizedYoutubeVideo[] => {
+  return videos.map((video) => {
+    return {
+      ...video,
+      normalized_score: {
+        date: normalize(video.raw_score.date, 0, 10),
+        dateXLikes: normalize(
+          video.raw_score.dateXLikes,
+          0,
+          maxScores.dateXLikes
+        ),
+        dateXViews: normalize(
+          video.raw_score.dateXViews,
+          0,
+          maxScores.dateXViews
+        ),
+        useOfChapters: video.raw_score.useOfChapters,
+      },
+    };
+  });
+};
+
+export const getWeightedExternalRanking = (
+  videos: INormalizedYoutubeVideo[]
+): IWeightedYoutubeVideo[] => {
+  return videos.map((video) => {
+    const weighted_score = {
+      date: video.normalized_score.date * WEIGHTS.w1,
+      dateXLikes: video.normalized_score.dateXLikes * WEIGHTS.w2,
+      dateXViews: video.normalized_score.dateXViews * WEIGHTS.w3,
+      useOfChapters: video.normalized_score.useOfChapters * WEIGHTS.w5,
+    };
+
+    const final_score =
+      0.5 * (weighted_score.date + weighted_score.dateXLikes) +
+      0.3 * weighted_score.dateXViews +
+      0.2 * weighted_score.useOfChapters;
+
+    return {
+      ...video,
+      weighted_score,
+      final_score,
+    };
+  });
+};
+
+export const getMaxScores = (
+  rawExternalRankingVideos: IRawYoutubeVideo[]
+): IMaxScores => {
+  const maxScore = {
+    dateXViews: 0,
+    dateXLikes: 0,
+  };
+
+  for (const video of rawExternalRankingVideos) {
+    maxScore.dateXViews = Math.max(
+      maxScore.dateXViews,
+      video.raw_score.dateXViews
+    );
+    maxScore.dateXLikes = Math.max(
+      maxScore.dateXLikes,
+      video.raw_score.dateXLikes
+    );
+  }
+
+  return maxScore;
+};
+
+export const getDateScore = (yearsSincePublished: number) => {
+  // Following the function in https://docs.google.com/document/d/1zxYRyytmbbvfAZkQc8dampD9Vhun0XQtWepKYxWUTKo
+  return Math.max(-Math.pow(1.6, yearsSincePublished) + 11, 0);
+};
+
+export const getDateXViewsScore = (
+  views: number,
+  daysSincePublished: number
+) => {
+  return views / daysSincePublished;
+};
+
+export const getDateXLikes = (likes: number, daysSincePublished: number) => {
+  return likes / daysSincePublished;
+};
+
+export const getUseOfChapters = (description: string) => {
+  const usesChapters = description.match("[0-9]:[0-9]");
+  return usesChapters !== null ? 1 : 0;
+};
+
+const getChannelPopularity = () => {
+  return 1;
+};
+
+export const getDaysSincePublished = (publishedAt: string) => {
+  const publishedAtDate = parseISO(publishedAt);
+  const daysSincePublished = differenceInCalendarDays(
+    new Date(),
+    publishedAtDate
+  );
+  return daysSincePublished;
+};
+
+export const ensureWeightsAreCorrect = (weights: typeof WEIGHTS) => {
```

Comment:
should we move this function to a test? I feel the only way this weights would be wrong is if an eng changed by mistake and a test can catch that before pushing to prod

Reply:
actually you are already testing this! so I would remove from here, wdyt?